### PR TITLE
[Merged by Bors] - Fix Events::<drain/clear> bug

### PR DESCRIPTION
**Pull request by NathanSWard**
Taken from #2145

On draining and clearing, dangling `EventReaders` would not read into the correct event offset.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thanks! This is a nice thing to pull over.

**Comment by cart** [comment]
Good call. Looks like `events.clear()` also needs to reset its counters. I'll put together a follow up.

**Comment by cart** [comment]
Hmm one issue to discuss that I discovered when putting together my followup:

Resetting `event_count` means that existing event readers will skip over newly sent events? The test looks like it should fail, so I need to understand why before merging :)

**Comment by cart** [comment]
image

**Comment by NathanSWard** [comment]
> Good call. Looks like events.clear() also needs to reset its counters. I'll put together a follow up.

@cart do you want me to add this into this PR?

> Resetting event_count means that existing event readers will skip over newly sent events? The test looks like it should fail, so I need to understand why it works before merging :)

Hmm yes this does seem to be incorrect...

**Comment by cart** [comment]
This makes the test above pass, but it feels like I'm missing a corner case:
image

**Comment by NathanSWard** [comment]
Yep adding this:
```
        events.send(E(1));
        let _ = events.drain();
        assert_eq!(reader.iter(&events).next(), None);
```
in the tests causes it to fail.

**Comment by cart** [comment]
> Yep adding this:

Is that with or without the "fix" above? Adding that passes with the fix for me.

**Comment by NathanSWard** [comment]
@cart, a possible solution: 
we might need to add a `version` field to `Events` and `Event<Reader/Writer>` and on calling clear/drain, update the version. So when the Reader/Writer calls `.iter()` check the versions and reset the Reader/Writer's internal count to `0` if the version has changed.

**Comment by NathanSWard** [comment]
> Is that with or without the "fix" above? Adding that passes with the fix for me.

With your change included.
here is the full test i have:
```rust
#[test]
    fn test_events_drain() {
        #[derive(PartialEq, Eq, Debug)]
        struct E(usize);

        let mut events = Events::<E>::default();
        let mut reader = events.get_reader();

        assert!(reader.iter(&events).next().is_none());

        events.send(E(0));
        assert_eq!(reader.iter(&events).next(), Some(&E(0)));
        assert_eq!(reader.iter(&events).next(), None);

        events.send(E(1));
        let _ = events.drain();
        assert_eq!(reader.iter(&events).next(), None);

        events.send(E(2));
        events.send(E(3));
        assert_eq!(reader.iter(&events).next(), Some(&E(2)));
        assert_eq!(reader.iter(&events).next(), Some(&E(3)));
        assert_eq!(reader.iter(&events).next(), None);
    }
```

**Comment by cart** [comment]
Found another corner case. Adding an `events.update()` breaks it.

```rust
#[test]
fn test_events_drain() {
    #[derive(PartialEq, Eq, Debug)]
    struct E(usize);
    let mut events = Events::<E>::default();
    let mut reader = events.get_reader();

    assert!(reader.iter(&events).next().is_none());

    events.send(E(0));
    assert_eq!(*reader.iter(&events).next().unwrap(), E(0));
    assert_eq!(reader.iter(&events).next(), None);

    let _ = events.drain();
    assert!(reader.iter(&events).next().is_none());

    events.send(E(1));
    events.update();
    events.send(E(2));
    assert_eq!(*reader.iter(&events).next().unwrap(), E(1));
    assert_eq!(*reader.iter(&events).next().unwrap(), E(2));
}
```

**Comment by cart** [comment]
I'm gonna keep working through my PR review queue. If you find a fix definitely do it for clear() too :smile:

I'll check back in later / help out if it hasn't been sorted out yet.

**Comment by NathanSWard** [comment]
@cart it turns out our last `#[test]`s are actually invalid... haha

`reader.iter(&events)` returns an iterator over all events. 
We were calling it multiple times expecting it to return a single value.

This is how the test should be structured...
```rust
#[test]
    fn test_events_drain() {
        #[derive(PartialEq, Eq, Debug)]
        struct E(usize);
        let mut events = Events::<E>::default();
        let mut reader = events.get_reader();

        assert!(reader.iter(&events).next().is_none());

        events.send(E(0));
        assert_eq!(*reader.iter(&events).next().unwrap(), E(0));
        assert_eq!(reader.iter(&events).next(), None);

        let _ = events.drain();
        assert!(reader.iter(&events).next().is_none());

        events.send(E(1));
        events.update();
        events.send(E(2));

        let mut iter = reader.iter(&events);
        assert_eq!(*iter.next().unwrap(), E(1));
        assert_eq!(*iter.next().unwrap(), E(2));
    }
```

**Comment by cart** [comment]
Bahaha whoops. Nice catch!

**Review by alice-i-cecile — APPROVED** [review]
Good; factoring out that logic reduces the risk of bugs in the future.

**Comment by NathanSWard** [comment]
@cart I have this current implementation which passed the failing test we set up.
I'm currently toying around with other potential failing tests, and please let me know if you can think of any :)

File: crates/bevy_ecs/src/event.rs [hunk 1]
```diff
@@ -328,14 +328,25 @@ impl<T: Component> Events<T> {
         events.update();
     }
 
+    #[inline]
+    fn increment_start_event_count(&mut self) {
+        let count = self.event_count + 1;
+        self.a_start_event_count = count;
+        self.b_start_event_count = count;
```
Comment (cart) [inline crates/bevy_ecs/src/event.rs:335]:
```suggestion
        self.a_start_event_count = self.event_count;
        self.b_start_event_count = self.event_count;
```

Just realized we can just use `event_count` directly. `event_count` corresponds to the "next" event, not the current event.

`event_count + 1` _probably_ also works due to how we reset event readers, but I'd rather not roll the dice :)
